plugins/outputs/postgresql/postgresql.go

```diff
@@ -0,0 +1,424 @@
+	for {
+		select {
+		case tableSource, ok := <-p.writeChan:
+			if !ok {
+				return
+			}
+			if err := p.writeRetry(ctx, tableSource); err != nil {
+				p.Logger.Errorf("write error (permanent, dropping sub-batch): %v", err)
+			}
+		case <-p.dbContext.Done():
+			return
+		}
+	}
+}
+
+// This is a subset of net.Error
+type maybeTempError interface {
+	error
+	Temporary() bool
+}
+
+// isTempError reports whether the error received during a metric write operation is temporary or permanent.
+// A temporary error is one that if the write were retried at a later time, that it might succeed.
+// Note however that this applies to the transaction as a whole, not the individual operation. Meaning for example a
+// write might come in that needs a new table created, but another worker already created the table in between when we
+// checked for it, and tried to create it. In this case, the operation error is permanent, as we can try `CREATE TABLE`
+// again and it will still fail. But if we retry the transaction from scratch, when we perform the table check we'll see
+// it exists, so we consider the error temporary.
+func isTempError(err error) bool {
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr); pgErr != nil {
+		// https://www.postgresql.org/docs/12/errcodes-appendix.html
+		errClass := pgErr.Code[:2]
+		switch errClass {
+		case "42": // Syntax Error or Access Rule Violation
+			switch pgErr.Code {
+			case "42701": // duplicate_column
+				return true
+			case "42P07": // duplicate_table
+				return true
+			}
+		case "53": // Insufficient Resources
+			return true
+		case "57": // Operator Intervention
+			return true
+		case "23": // Integrity Constraint Violation
+			switch pgErr.Code {
+			case "23505": // unique_violation
+				if strings.Contains(err.Error(), "pg_type_typname_nsp_index") {
+					// Happens when you try to create 2 tables simultaneously.
+					return true
+				}
+			}
+		}
+		// Assume that any other error that comes from postgres is a permanent error
+		return false
+	}
+
+	if mtErr := maybeTempError(nil); errors.As(err, &mtErr) {
+		return mtErr.Temporary()
+	}
+
+	// Assume that any other error is permanent.
+	// This may mean that we incorrectly discard data that could have been retried, but the alternative is that we get
+	// stuck retrying data that will never succeed, causing good data to be dropped because the buffer fills up.
+	return false
+}
+
+func (p *Postgresql) writeRetry(ctx context.Context, tableSource *TableSource) error {
+	backoff := time.Duration(0)
+	for {
+		tx, err := p.db.Begin(ctx)
+		if err != nil {
+			return err
+		}
+
+		err = p.writeMetricsFromMeasure(ctx, tx, tableSource)
+		if err == nil {
+			tx.Commit(ctx)
+			return nil
+		}
+
+		tx.Rollback(ctx)
+		if !isTempError(err) {
+			return err
+		}
+		p.Logger.Errorf("write error (retry in %s): %v", backoff, err)
+		tableSource.Reset()
+		time.Sleep(backoff)
+
+		if backoff == 0 {
+			backoff = time.Millisecond * 250
```

Comment:
You might want to initialize the `backoff` with this value to simplify the code.

Reply:
The initial backoff is deliberately 0. Meaning the first retry is immediate.

Reply:
Right, that's how I understood it, but if you can live with an initial delay between the retries, you can simplify the code.

Reply:
I would prefer the initial retry not have an artificial delay. I don't think the code is complex, but if you prefer, an exponential backoff could be used instead, which would calculate to an initial delay of 0. I just preferred a doubling backoff as exponential climbs too fast. But I would prefer that over an initial delay.